File: ninanatur/garden/roofs.py
"""What a roof does to the shadow a building casts.

OSM's `height` is the **ridge**, so a building modelled as a prism at that height
is a building whose gable ends are solid to the ridge. It shades too much,
everywhere, all season — and the taller the roof, the further the error reaches.

A roof shape is one of the few things somebody can answer by looking out of the
window, which is the test this project applies to every question it asks. So it
is asked, and where it is not answered the model says what it assumed.
"""
from __future__ import annotations

from enum import StrEnum


class Roof(StrEnum):
    """The shapes a house in Germany actually has.

    Stored as text, so a renamed member orphans saved rows — the values are part
    of the data.
    """

    FLAT = "flat"
    GABLE = "gable"
    HIP = "hip"
    PENT = "pent"
    #: Nobody has said. The model then treats the recorded height as solid,
    #: which is what it did for every building before this existed.
    UNKNOWN = "unknown"


#: How much of the roof's own rise still shades, as a fraction.
#:
#: A flat roof is the whole building at one height. A gable is a wedge: the
#: ridge is a line and the eaves are the edges, so averaged across the footprint
#: about half of the rise stands. A hip slopes on all four sides and stands for
#: less. A pent is a single slope and keeps more than a gable, because one whole
#: edge is at full height.
#:
#: These are ratios of a shape, not measurements of a building. They are here
#: rather than inline so that somebody who disagrees has one place to argue.
RISE_KEPT: dict[Roof, float] = {
    Roof.FLAT: 1.0,
    Roof.GABLE: 0.5,
    Roof.HIP: 0.4,
    Roof.PENT: 0.6,
    Roof.UNKNOWN: 1.0,
}

#: Where the eaves are when nothing says: three quarters of the way up.
#:
#: A two-storey German house is roughly 6 m to the eaves and 8 to the ridge.
#: `building:levels x 3 m` is the better answer wherever OSM carries it, and
#: this is what stands in when it does not — an assumption, and labelled as one
#: beside the assumed heights that are already there.
DEFAULT_EAVES_FRACTION = 0.75


def shading_height(
    height_m: float, roof: Roof = Roof.UNKNOWN, eaves_m: float | None = None
) -> float:
    """The height to shade with, given the ridge height and the roof.

    Never above the recorded height and never below the eaves: a roof takes
    something off the top and cannot take off the building.
    """
    if roof is Roof.UNKNOWN or roof is Roof.FLAT:
        return height_m
    eaves = DEFAULT_EAVES_FRACTION * height_m if eaves_m is None else eaves_m
    eaves = max(0.0, min(eaves, height_m))
    return eaves + RISE_KEPT[roof] * (height_m - eaves)


def eaves_from_levels(levels: float | None, storey_m: float = 3.0) -> float | None:
    """Eaves height from `building:levels`, or None.

    Floor to floor, which is what a storey count means. It is the one input here
    that is a measurement rather than a shape ratio, so it wins wherever OSM
    carries it.
    """
    if levels is None or levels <= 0:
        return None
    return levels * storey_m

File: frontend/src/roofs.ts
/**
 * The roof shapes the shading model has a ratio for.
 *
 * Mirrors `ninanatur/garden/roofs.py::Roof`, and a pytest guard fails if the
 * two fall out of step (`tests/test_kind_vocabulary.py`, since Wave 21 — before
 * that this said so and nothing checked) — the kind vocabulary learnt that
 * lesson once already, when a dropdown went on offering a value the server had
 * stopped knowing.
 *
 * Asked because OSM's `height` is the ridge: without a shape a house is
 * modelled as solid to it, and shades too much all season. It is also one of
 * the few things somebody can answer by looking out of the window.
 */
export const ROOFS: ReadonlyArray<readonly [string, string]> = [
  ['unknown', 'weiß nicht'],
  ['gable', 'Satteldach'],
  ['hip', 'Walmdach'],
  ['pent', 'Pultdach'],
  ['flat', 'Flachdach'],
  // Surveyed answers, added in Wave 19 when the official models turned out to
  // give them. A fifth of German buildings are Mischform, and until now
  // somebody with one had to lie to this list.
  ['mix', 'Mischform'],
  ['other', 'andere Form'],
];

/**
 * Where the eaves are when nobody has said: three quarters of the way up.
 *
 * Mirrors `ninanatur/garden/roofs.py::DEFAULT_EAVES_FRACTION`, and a pytest
 * guard holds the two together — the page says what the model assumed (doc 93),
 * so it has to assume the same thing.
 */
export const EAVES_FRACTION = 0.75;

/** Kinds a roof is a sensible question for. A pond has no roof. */
export const ROOFED = new Set(['house', 'shed']);
